src/ice/agent.rs

```diff
@@ -761,63 +761,82 @@ impl IceAgent {
     /// Tells whether the message is for this agent instance.
     ///
     /// This is used to multiplex multiple ice agents on a server sharing the same UDP socket.
-    /// For this to work, the server should operate in ice-lite mode and not initiate any
-    /// binding requests itself.
     ///
-    /// If no remote credentials have been set using `set_remote_credentials`, the remote
-    /// ufrag is not checked.
+    /// For binding requests, if no remote credentials have been set using
+    /// `set_remote_credentials`, the remote ufrag is not checked.
     pub fn accepts_message(&self, message: &StunMessage<'_>) -> bool {
         trace!("Check if accepts message: {:?}", message);
 
-        // The username for the credential is formed by concatenating the
-        // username fragment provided by the peer with the username fragment of
-        // the ICE agent sending the request, separated by a colon (":").
-        if message.is_binding_request() {
-            // The existence of USERNAME is checked in the STUN parser.
-            let (local, remote) = message.split_username().unwrap();
-
-            let local_creds = self.local_credentials();
-            if local != local_creds.ufrag {
-                trace!(
-                    "Message rejected, local user mismatch: {} != {}",
-                    local,
-                    local_creds.ufrag
-                );
-                return false;
+        let do_integrity_check = |is_request: bool| -> bool {
+            let (_, password) = self.stun_credentials(is_request);
+            let integrity_passed = message.check_integrity(&password);
+
+            // The integrity is always the last thing we check
+            if integrity_passed {
+                trace!("Message accepted");
+            } else {
+                trace!("Message rejected, integrity check failed");
             }
+            integrity_passed
+        };
 
-            if let Some(remote_creds) = &self.remote_credentials {
-                if remote != remote_creds.ufrag {
+        let method = message.method();
+        let class = message.class();
+        match (method, class) {
+            (StunMethod::Binding, StunClass::Request | StunClass::Indication) => {
```

Comment:
RFC 5389 indicates that requests and indications should be [validated in a similar manner](https://www.rfc-editor.org/rfc/rfc5389#section-10.1.2).